libs/tracing-utils/src/http.rs

```diff
@@ -0,0 +1,95 @@
+//!
+//! Tracing wrapper for Hyper HTTP server
+//!
+
+use hyper::HeaderMap;
+use hyper::{Body, Request, Response};
+use std::future::Future;
+use tracing_opentelemetry::OpenTelemetrySpanExt;
+
+/// Configuration option for what to use as the "otel.name" field in the traces.
+pub enum OtelName<'a> {
+    /// Use a constant string
+    Constant(&'a str),
+
+    /// Use the path from the request.
+    ///
+    /// That's very useful information, but is often left out by default because it
+    /// might contain sensitive information like usernames or email addresses.
+    UriPath,
+}
+
+/// Handle an incoming HTTP request using the given handler function,
+/// with OpenTelemetry tracing.
+///
+/// This runs 'handler' on the request in a new span, with fields filled in
+/// from the request. Notably, if the request contains tracing information,
+/// it is propagated to the span, so that this request is traced as part of
+/// the same trace.
+///
+/// XXX: Usually, this is handled by existing libraries, or built
+/// directly into HTTP servers. However, I couldn't find one for Hyper,
+/// so I had to write our own. OpenTelemetry website has a registry of
+/// instrumentation libraries at:
+/// https://opentelemetry.io/registry/?language=rust&component=instrumentation
+/// If a Hyper crate appears, consider switching to that.
+pub async fn tracing_handler<F, R>(
+    req: Request<Body>,
+    handler: F,
+    otel_name: OtelName<'_>,
+) -> Response<Body>
+where
+    F: Fn(Request<Body>) -> R,
+    R: Future<Output = Response<Body>>,
+{
+    // Start a tracing span, with context propagated from the incoming request
+    // request if any.
+    //
+    // See list of standard fields defined for HTTP requests at
+    // https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/http.md
+    // We only fill in a few of the most useful ones here.
+    let otel_name = match otel_name {
+        OtelName::Constant(s) => s,
+        OtelName::UriPath => req.uri().path(),
+    };
+
+    let span = tracing::info_span!(
+        "http request",
+        otel.name= %otel_name,
+        http.method = %req.method(),
+        http.status_code = tracing::field::Empty,
+    );
+    let parent_ctx = extract_remote_context(req.headers());
+    span.set_parent(parent_ctx);
+    let _ = span.enter();
+
+    // Handle the request
+    let response = handler(req).await;
```

Comment:
This seems wrong, instead with futures the instrument wrapping future to enter/exit with each poll should be the way to go:

```suggestion
    // Handle the request
    let response = handler(req).await.instrument(span);
```

Reply:
Also, should we catch_unwind here?

Reply:
I see you changed this. I'd double check with jaeger that the span cloning doesn't create any surprising effects on the traces..

Reply:
Yep, tested it.

`Span::current()` also returns a clone of the Span, rather than `&Span`. I was surprised. A Span is apparently just a reference to the underlying "span thingie", like an Arc. Although looking at the implementation of `Span::clone`, it's a little more complicated than incrementing a reference count.

Reply:
Good catch on the original issue, thanks!